examples/pwm-input.rs

```diff
@@ -0,0 +1,48 @@
+#![deny(unsafe_code)]
+#![no_main]
+#![no_std]
+
+// Halt on panic
+use panic_halt as _;
+
+use cortex_m;
+use cortex_m_rt::entry;
+use stm32f4xx_hal::{prelude::*, pwm, pwm_input, stm32, timer};
+
+#[entry]
+fn main() -> ! {
+    if let Some(dp) = stm32::Peripherals::take() {
+        // Set up the system clock.
+        let rcc = dp.RCC.constrain();
+        let clocks = rcc.cfgr.freeze();
+
+        let gpioa = dp.GPIOA.split();
+        let gpioc = dp.GPIOC.split();
+
+        let channels = (
+            gpioa.pa8.into_alternate_af1(),
+            gpioa.pa9.into_alternate_af1(),
+        );
+        // configure tim1 as a PWM output of known frequency.
+        let pwm = pwm::tim1(dp.TIM1, channels, clocks, 501u32.hz());
+        let (mut ch1, _ch2) = pwm;
+        let max_duty = ch1.get_max_duty();
+        ch1.set_duty(max_duty / 2);
+        ch1.enable();
+        // configure tim8 as a timer, using the best-guess frequency of the input signal.
+        // FIXME: using this for the side effect of configuring the timer, do this ourselves.
+        let tim8 = timer::Timer::tim8(dp.TIM8, 500.hz(), clocks).release();
+
+        // Configure tim8's channel 1 pin into AF3
+        let pwm_reader_ch1 = gpioc.pc6.into_alternate_af3();
+        // instantiate the PWM reader from tim8 and tim8 channel 1.
+
+        let pwm_reader = pwm_input::PwmInput::tim8(tim8, pwm_reader_ch1);
```

Comment:
I plan to change timers API to be similar to `stm32f1xx-hal`. See #337.
With this change you could rewrite it to:
```rust
let pwm_reader_ch1 = gpioc.pc6.into_alternate();
let pwm_reader = timer::Timer:new(dp.TIM8, clocks).pwm_input(pwm_reader_ch1);
```

Reply:
Thanks. Ended up rewriting this PR due to #337 but the result is much cleaner this time around.